### PR TITLE
Improve takeWithin precision

### PR DESCRIPTION
**Pull request by johanandren**
## Purpose
Improve takeWithin precision so it is less likely to let elements through after the timeout when a tight timeout is used.

## Changes
In addition to completing the stage on a scheduled timer event, which may be delayed because of
the graph interpreter being busy with other stages, we also check for each push and pull if the timeout
has hit already and complete the stage immediately in that case.

### CODE REVIEW

**Review by johanandren — COMMENTED** [review]
Couldn't come up with a good way to cover this with test since it is so time sensitive.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -1846,9 +1846,18 @@ private[stream] object Collect {
 
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
     new TimerGraphStageLogic(shape) with InHandler with OutHandler {
-      def onPush(): Unit = push(out, grab(in))
+      private val timeoutNs = System.nanoTime() + timeout.toNanos
 
-      def onPull(): Unit = pull(in)
+      def overdue() = System.nanoTime() >= timeoutNs
```
Comment (hepin1989) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Then there will be so much system call.

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Yeah, that's the downside of it. One guard per element. Perhaps we could toggle it and only use it when the timeout is very short (which should anyway limit the number of elements).

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 2]
```diff
@@ -1725,43 +1725,50 @@ private[stream] object Collect {
...
+              failStage(BufferOverflowException(s"Buffer overflow for delay operator (max capacity was: $size)!"))
+            }
         case _: Backpressure =>
-          () => {
-            throw new IllegalStateException("Delay buffer must never overflow in Backpressure mode")
-          }
+          () =>
+            {
+              throw new IllegalStateException("Delay buffer must never overflow in Backpressure mode")
+            }
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: I have no idea what ScalaFMT is up to, formats differently in IntelliJ on save than from SBT now ;(

**Review by helena — APPROVED** [review]
Nice. Yes there's the downside, though this is better, and the upside.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 1 above)
Comment (hepin1989) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: The new way is OK:)

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 3]
```diff
@@ -1844,11 +1844,24 @@ private[stream] object Collect {
  */
 @InternalApi private[akka] final class TakeWithin[T](val timeout: FiniteDuration) extends SimpleLinearGraphStage[T] {
 
+  // avoid excessive system calls for longer timeouts
+  private val shortTimeout = timeout < 1.second
```
Comment (hepin1989) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: great!

Comment (hepin1989) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: how about changing the 1.seconds a const value `threshold`?

**Review by jrudolph — COMMENTED** [review]
Interesting trade-off here. The underlying assumption seems to be that expected starvation in the streams infrastructure should be shorter than one second.

On the other hand, I wonder how big an issue precision is at all. You weren't able to write a test but it might be impossible to write a test because it has an inherent race condition. Whatever precision we have here, you cannot guarantee whether a downstream stage will see elements after the deadline because there can always be elements "in the pipe" (more precisely, at most one element can be in the pipe to the immediately following fused stage). So, if we check the deadline for every element, we can still only guarantee that the immediately following stage will see at most one more element after the deadline. Is that the improvement we are looking for?

One way to make it cheaper would be, not to use the regular timer logic and time check here but use the system scheduler to toggle a `@volatile var` which is checked on every push (and maybe pull).

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 4]
```diff
@@ -1839,16 +1839,36 @@ private[stream] object Collect {
...
-      def onPull(): Unit = pull(in)
+      private def overdue() = shortTimeout && System.nanoTime() >= timeoutNs
+
+      def onPush(): Unit = {
+        if (overdue()) completeStage()
+        else push(out, grab(in))
+      }
+
+      def onPull(): Unit = {
+        if (overdue()) completeStage()
```
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: The only backside of that is that it increases the risk of elements being dropped, but there's always that risk so ¯\_(ツ)_/¯

**Comment by johanandren** [comment]
@jrudolph Yeah, not quite sure this PR is worth it at all. 

Choosing one second was not about that the starvation or unfairness will never be longer but rather that the granularity is mostly interesting for small numbers (vs `takeWhile(1 second and 37 millis)`) and that a longer time span will likely see more elements coming through which would all pay the syscall price.

**Pull request closed without merging**